### ISSUE TITLE
PV can not be reconnected after disconnect

### ISSUE THREAD

**Issue by protofabrik** [pyepics/pyepics#68]
We've stumbled across a rather inconvenient bug, that is after we disconnect from the PV the PV can not be reconnected anymore. Neither PV.connect nor PV.force_connect work correctly. The first one never connects, the second one successfully connects to the channel but fails with key error when re-adding the callback. See attached files for stand-alone demonstration of the bug... 

``` python
"""
Test-case that demonstrates that after disconnect and channel can not be reconnected using connect method

To try it out, simply start example CAS server bundled with EPICS base (excas binary) before starting the test

The output for me is: 

Connecting
UPD 6.63983507548
SUM 6.63983507548
UPD 76.9632902148
Disconnecting
Reconnecting
Traceback (most recent call last):
  File "test.py", line 33, in <module>
    assert(pv_test.wait_for_connection(5.0) == True)
AssertionError

"""

import epics
import time

pv_test = epics.PV("alan",auto_monitor=True)


def callback(value=None,**kwargs):
    print("UPD",value.sum())



pv_test.add_callback(callback)

print("Connecting")
assert(pv_test.wait_for_connection(5.0) == True)

print("SUM",pv_test.get().sum())

time.sleep(3)

print("Disconnecting")
pv_test.disconnect()
print("Reconnecting")

# Always fails :( 
pv_test.connect()
assert(pv_test.wait_for_connection(5.0) == True)
pv_test.add_callback(callback)


time.sleep(5)
```

``` python
"""
Test-case that demonstrates that after disconnect a channel can be
reconnected using undocumented force_connect, however the callback can not be added again (after force_connect no original callbacks are invoked)

To try it out, simply start example CAS server bundled with EPICS base (excas binary) before starting the test

The output for me is: 

Connecting
UPD 6.92243429227
SUM 6.92243429227
UPD -6.11459633615
Disconnecting
Reconnecting
Traceback (most recent call last):
  File "test2.py", line 49, in <module>
UPD -6.11459633615
    pv_test.add_callback(callback)
  File "/usr/lib/python3.5/site-packages/epics/pv.py", line 502, in add_callback
    self.get_ctrlvars()
  File "/usr/lib/python3.5/site-packages/epics/pv.py", line 425, in get_ctrlvars
    kwds = ca.get_ctrlvars(self.chid, timeout=timeout, warn=warn)
  File "/usr/lib/python3.5/site-packages/epics/ca.py", line 419, in wrapper
    return fcn(*args, **kwds)
  File "/usr/lib/python3.5/site-packages/epics/ca.py", line 1418, in get_ctrlvars
    ncache = _cache[current_context()][name(chid)]
KeyError: 'alan'


"""

import epics
import time

pv_test = epics.PV("alan",auto_monitor=True)


def callback(value=None,**kwargs):
    print("UPD",value.sum())



pv_test.add_callback(callback)

print("Connecting")
assert(pv_test.wait_for_connection(5.0) == True)

print("SUM",pv_test.get().sum())

time.sleep(3)

print("Disconnecting")
pv_test.disconnect()
print("Reconnecting")

# Always fails :( 
pv_test.force_connect()
assert(pv_test.wait_for_connection(5.0) == True)
pv_test.add_callback(callback)


time.sleep(5)
```

**Comment by newville** [comment]
@protofabrik   Hm,  I see slightly different (but hard to call "correct"!) behavior, in that using `PV.force_reconnect()` does actually reconnect for me, but the PV is in some weird state as the global cache has not been fully cleared and repopulated.   

We should try to fix this, but I think that if #69 is the pressing problem , the solution to that problem is somewhat simpler.

**Comment by jsparger** [comment]
Has there been any progress on this? I am also having to work around this problem.

**Comment by newville** [comment]
@jsparger @protofabrik  I've played around with this a while back, but didn't come up with a clean solution.     Again, I still see the problem as what I described earlier:  PV.force_reconnect() works, but leaves the PV in sort of a funky state because it has been cleared from the global caches (`ca._cache` and `pv._PVcache_`).

Then again, I manage to not get in the situation where I want to disconnect and then re-connect to a PV.  In fact, I would be hard-pressed to find a PV.disconnect() in any of my codes.   That's not to say that I don't think this behavior is not a problem -- I think it is something you should be able to do.  But it is to say that I'm not 100% sure that I know what `disconnect()` and `reconnect()` should do with the global caches. 

Maybe it's worth asking what that use case is: What is disconnecting supposed to do?  What is reconnecting supposed to do?

**Comment by klauer** [comment]
A PV instance should not have to care about the existence of the global cache, as it's an implementation detail of pyepics. The desired functionality from the `epics.PV` methods are then rather clear with regard to`disconnect` and `connect`.

**Comment by newville** [comment]
@klauer Actually, the desired behavior of `disconnect` and `reconnect` are not clear me, and certainly not outside of the context of the implementation details of the global caches.   Please explain what you expect `PV.disconnect` should do.

**Comment by klauer** [comment]
`connect()` - I want to use this PV, get monitor callbacks, and keep its value updated internally

`disconnect()` - I don't want to use this PV any more, do not want to get monitor callbacks, do not want it to keep its value updated internally

Where is the confusion coming in?

**Comment by newville** [comment]
@klauer  well, there are a lot of small details, most of them exactly about the implementation detail of the global caches.

If the goal is to not get monitor callbacks, you can do `PV.clear_callbacks()`.   If you're worried about the automatic internal monitoring , that can be handled, though only at the ca level, and not very easily.    Also, if you're worried about the internal monitoring, can you demonstrate that it's a problem?

You could also just do `del pv`.   That will run the current version of `PV.disconnect()`  (clearing monitor callbacks, emptying the caches) and then throw away the reference.    You "reconnect" by making a new PV.   Is that not acceptable?  It is somehow "too expensive"?  I don't know.

Are these the "workarounds" that people have to do because "PV.disconnect() / reconnect()"?  Really, I don't know.

`PV.disconnect()` should apparently be different from those two options, so that `PV.reconnect()` "works" (for some definition of "works").   It seems that the goal of `PV.disconnect()` must be to retain *some* state, perhaps so that there is some efficiency to be had by re-connecting a disconnected PV (though that may be an illusion), or perhaps so that internal settings like the callback functions are preserved so that if you assign a callback, disconnect, and then reconnect, that callback will reactivate.   That is, it must be different from "delete then recreate", right?  So, again:  what state should be preserved, what part of the initial connection process can be skipped when reconnecting?  

Can one disconnect in one thread and reconnect in another? 

This is why I asked for what it should do, and why it seems to me that it is definitely about the the global caches.   I'm not saying that the current `PV.disconnect()` /  `reconnect()` are acceptable -- there is clearly a problem.  I just don't know what the right behavior should be.

**Comment by tacaswell** [comment]
I think that `connect` should make to `ca.create_channel` and `disconnect` should map to `ca.clear_channel`, along with what ever book keeping needs to be done on the python side.

I suspect that `_PVcache_` should be done with a weakref dict and only used with `get_pv` (not direct calls to `PV`) which gets that management out of `disconnect`.

If I am understanding the `ca.py` code correctly, the `ca._cache` will take care of it's self via `ca._onConnectionEvent`.

You probably _should_ be able to disconnect on one thread and reconnect on another, however I can not find any thread locking primatives in the code (other than what we get from the gil for free) so I would be surprised if pyepics was thread safe in general.

I'll take a crack at this.

**Comment by newville** [comment]
@tacaswell 

PV initialization does a fair amount more than `ca.create_channel`, including setting up a connection callback that then does more setting of state information, generally so that PV.connect() does not need to be called -- we definitely do not want to give that up.  The global caching of connection state is not merely convenience: the caching of callbacks is required to prevent them from being garbage collected.   `ca._cache` cannot use weak references.

The PV cache is mostly just convenience so that `get_pv()` on the same channel does not have to create a new connection.   This is a common usage pattern (say, with 'caget()' and 'caput()'), and we definitely want to keep these.  So, I think they should not use weakrefs either -- the point is to keep them around for later use -- but maybe I'm not understanding your point.  

PV.disconnect() could do ca.clear_channel(), but that will not cause the connection callback to run.   The client side does not trigger those events, they happen from the CA library when the client loses or regains connection to the server.   So doing ca.clear_channel will not automatically clear the cache entry, but it will cause the connection to not work, and may cause core-dumps on exit unless the ca._cache is cleared correctly.

Threads with CA are slightly complicated, but CA and pyepics can be thread-safe with some care.   Basically, one has to make sure that PVs are created and used with one CA context, which is a per-thread thing.  This means either that PVs are used within an individual thread, or yo have to explicitly set the CA context to allow PVs to cross threads.  The CAThread class creates a Thread and attaching to the initial CA context created on when globally initializing the library, based on the recommendation from the Epics gurus.  For the issue here, it may be a good idea to preserve the channels context (and so at least some of the information in ca._cache) on PV.disconnect().

But again, I think the fundamental question here is what global state information PV.disconnect should preserve.   If the answer is "nothing", than "del pv" is what you want.  

So, really: what are the use cases for wanting to disconnect and then reconnect?

**Comment by tacaswell** [comment]
Preserving the python-side callbacks seems like a good idea.   There is also details about if the returned value for waveforms should be a list or string, if it should be text, what the length should be, automonitor,  etc.   Some of that needs to get baked in when you connect as I understand it (would have to dig into it to tell you exactly which ones) so the ability to get a PV object in, disconnect it, change some settings, and then reconnect is very useful (we currently have some dodgy code on ophyd to deal with this for example).

It would also be nice to have the option of not connecting as soon as the PV is created, but that is a bit tangential to this.

I have a working version of the `_PVCache_` changes, however it is not useful without the rest of the changes, which gave me more trouble than expected :smiling_imp: 

https://github.com/pyepics/pyepics/blob/master/epics/pv.py#L800 is what seems to remove the pv from clear `ca._cache`, what I meant was that the `_onConnect` will correctly re-populate that.

The part I ran into trouble on is that the `chid` is a c++ pointer, if you try to call `clear_channel` with it twice, it will segfault.  I think I know how to deal with this.

I think a bigger problem that this has brought up is that `clear_channel` is currently only called on exit which means even if you delete your pv, the c++ objects will live for the life of your process.

On the other hand, if you think it is not worth making these methods work correctly they should just be removed.

**Comment by newville** [comment]
@tacaswell  

> On the other hand, if you think it is not worth making these methods work correctly they 
> should just be removed.

For me it's more that (as I tried to say above) I don't actually know what `PV.disconnect()` + ` PV.reconnect()` should do to "work correctly".   How should this differ from deleting the PV and recreating it?   We can make choices and then implement those, but I'm not sure "correct" is well-defined.    Like, maybe `PV.disconnect()` should save the initialization options (auto_monitor, form, count, callbacks) and then just `PV.reconnect()` should just re-run `PV.__init__()` with those arguments.    But, maybe the callbacks should not be re-instated.   Again, it would be nice to hear expected use cases for `PV.disconnect()' + `PV.reconnect()`.   

I'm not sure why any changes to the caches would be necessary.

> I think a bigger problem that this has brought up is that clear_channel is currently only called 
> on exit which means even if you delete your pv, the c++ objects will live for the life of your 
> process.

Hm, is *that* the goal for `PV.disconnect()`?   Is the fact that the CA layer holds onto a reference to that channel causing problems?   What are they (and why haven't I seen them)?   But, I agree that If it is causing problems, then adding `ca.clear_channel()` to `PV.disconnect()` when it removes the entry from `ca._cache()` might help.   I think that `ca.clear_channel` should probably empty the cache entries itself anyway: if `ca.clear_channel` is run, the caches should definitely be empty so recreating that PV starts from scratch.  Then again, I'd treat the whole approach as experimental and suggest careful testing of whether this actually solves any problem you are actually having. 

Another approach would be to have `ca.clear_channel()` happen at `del pv`, but not at `PV.disconnect()`.   That is, `PV.disconnect()` could do *less* (removing from PV cache, cancel user-defined callbacks, leave in `ca._cache`) so that `PV.reconnect()` can work.  In this scenario, `del pv` could really remove the PV from all caches and clear the channel from the process, while just letting a PV go out of scope does nothing (the current behavior, keeping that PV in the caches).  

> It would also be nice to have the option of not connecting as soon as the PV is created, but
> that is a bit tangential to this.

That would be possible.  PV.__init__() is generally very fast, and connections (which are definitely slower) happen asynchronously in the background.  I don't I've ever wanted to create a PV object but not have it set up those connections in the background    What do you do with an unconnected PV except wait for it to connect?

FWIW, I realized that threading state for a PV (like whether re-connection happens in the same thread as disconnection or whatever else) is actually not an issue.  That is, long ago, we followed recommendations from epics gurus so that all PVs always connect to the initial CA context (presumably created in the main thread or, for in a thread that will live for all CA usage in the process) and so should work from all threads.

**Comment by hhslepicka** [comment]
@newville I am getting hit by this effect of trying to `get_pv` with a previously pv that was disconnected and it just does not work... Do you have a recommended workaround?

**Comment by newville** [comment]
@hhslepicka I could wrong, but I thought this was sort of resolved, or at least that there was a good solution from @klauer, so that `PV.reconnect()` works:

    import epics
    import time
    pvname = 'Py:ao1'
    pv_test = epics.PV(pvname,  auto_monitor=True)

    def callback(value=None,  **kwargs):
        print("New Value: ", value)

    pv_test.add_callback(callback)
    print("connecting, wait for events:")
    time.sleep(5)

    print("disconnect:")
    pv_test.disconnect()
    time.sleep(5)

    print("reconnect:")
    pv_test.reconnect()

    pv_test.add_callback(callback)
    time.sleep(5)

I have to admit that I don't use `PV.disconnect()` in my own production code, so maybe I'm not seeing all the subtleties.  But I think it *should* work.  Does this work OK for you?  If not, can you give more details?

**Comment by hhslepicka** [comment]
@newville I'm really sorry. After banging my head trying to find the issue I tracked it down to my damaged local EPICS Base and caRepeater.
I nuked them out, started fresh and everything seems to be back to normal. I will run some more tests here and get back to you.

**Comment by newville** [comment]
@hhslepicka no worries!

**Comment by newville** [comment]
re-re-closing....
